tests-lit/tests/junk_detection/05_compilation_database_escaped_path/sample.cpp

```diff
@@ -0,0 +1,24 @@
+#include <string>
+#include <iostream>
+
+#ifndef ESCAPED_DEFINITION_STUB
+#error "ESCAPED_DEFINITION_STUB is not defined"
+#endif
+
+#define _STRINGIFY(a) #a
+#define STRINGIFY(a) _STRINGIFY(a)
+
+int main() {
+  bool equal = std::string(STRINGIFY(ESCAPED_DEFINITION_STUB)) == std::string("/src/builds/amd64-linux-mull");
```

Comment:
@chgans, thanks for following up on my progress. This is how it looks on my end:

If do this:

```
//#define _STRINGIFY(a) #a
//#define STRINGIFY(a) _STRINGIFY(a)

int main() {
  bool equal = std::string(ESCAPED_DEFINITION_STUB) == std::string("/src/builds/amd64-linux-mull");
  return ! equal;
}
```

I start getting this:

```
+ /opt/llvm-3.9.0/bin/clang++ -fembed-bitcode -g -O0 -DESCAPED_DEFINITION_STUB=/src/builds/amd64-linux-mull /sandbox/mull/tests-lit/tests/junk_detection/05_compilation_database_escaped_path/sample.cpp -o /sandbox/mull/tests-lit/tests/junk_detection/05_compilation_database_escaped_path/sample.cpp.exe
/sandbox/mull/tests-lit/tests/junk_detection/05_compilation_database_escaped_path/sample.cpp:12:28: error: expected expression
  bool equal = std::string(ESCAPED_DEFINITION_STUB) == std::string("/src/builds/amd64-linux-mull");
                           ^
<command line>:1:33: note: expanded from here
#define ESCAPED_DEFINITION_STUB /src/builds/amd64-linux-mull
                                ^
/sandbox/mull/tests-lit/tests/junk_detection/05_compilation_database_escaped_path/sample.cpp:12:28: error: use of undeclared identifier 'src'
```

Reply:
@chgans I will resolve this for now but you can also share which system are you on?

Reply:
We use a simple custom docker container, based on Ubuntu 18.04 amd64, with LLVM installed from apt.llvm.org.
Currently mull is build from source, using latest release. But i think I will eventually switch to a non-release version, as i would like to pick a few fixes you guys made recently (#659, and hopefully  #664 and this one #679).

I'm not familiar with mull source code, but i'm surprised you need a regexp to address the `\\\`, if you use a json parser, this should be solved for you. The same compilation DB is used with clang-tidy and clazy without any issue, and the macro evaluates to a C string.

I think it all boils down to how you execute the command line, eg do you pass it through a shell or you's exec'ing it?
If it goes through a shell, then the `"` need to be escaped so that the shell doesn't consume it, if you exec/fork it, then the `"` should not be escaped.

AFAIU, the JSON command property gives you a command line that should not (doesn't need to) be passed to a shell.

Reply:
Sorry, as per https://clang.llvm.org/docs/JSONCompilationDatabase.html#format
> Parameters use shell quoting and shell escaping of quotes, with ‘"’ and ‘\’ being the only special characters

The compile DB uses shell escaping, and this need to be undone because `clang::ASTUnit::LoadFromCommandLine()` doesn't execute the command in a shell.

AFAIU, the yaml/json loading deals with `\\\"` => `\"`, and you just need to undo the shell escapes by replacing any `\\` with `\` and any `\"`' with `"` before calling `clang::ASTUnit::LoadFromCommandLine`.

Reply:
@chgans, I have two solve two problems with this PR: your original report but also we need to set up integration tests to compile C++ fixtures which I am fighting with right now. I will get back to your comment above as soon as I get my current regex green.